cmd/kube-apiserver/app/server.go

```diff
@@ -95,34 +94,34 @@ cluster's shared state through which all other components interact.`,
 }
 
 // Run runs the specified APIServer.  This should never exit.
-func Run(commandOptions *options.ServerRunOptions) error {
-	kubeAPIServerConfig, sharedInformers, err := CreateKubeAPIServerConfig(commandOptions)
+func Run(runOptions *options.ServerRunOptions, stopCh <-chan struct{}) error {
+	kubeAPIServerConfig, sharedInformers, err := CreateKubeAPIServerConfig(runOptions)
 	if err != nil {
 		return err
 	}
-	kubeAPIServer, err := CreateKubeAPIServer(kubeAPIServerConfig, sharedInformers, wait.NeverStop)
+	kubeAPIServer, err := CreateKubeAPIServer(kubeAPIServerConfig, sharedInformers, stopCh)
 	if err != nil {
 		return err
 	}
 
 	// if we're starting up a hacked up version of this API server for a weird test case,
 	// just start the API server as is because clients don't get built correctly when you do this
 	if len(os.Getenv("KUBE_API_VERSIONS")) > 0 {
-		return kubeAPIServer.GenericAPIServer.PrepareRun().Run(wait.NeverStop)
+		return kubeAPIServer.GenericAPIServer.PrepareRun().Run(stopCh)
 	}
 
 	// otherwise go down the normal path of standing the aggregator up in front of the API server
 	// this wires up openapi
 	kubeAPIServer.GenericAPIServer.PrepareRun()
-	aggregatorConfig, err := createAggregatorConfig(*kubeAPIServerConfig.GenericConfig, commandOptions)
+	aggregatorConfig, err := createAggregatorConfig(*kubeAPIServerConfig.GenericConfig, runOptions)
 	if err != nil {
 		return err
 	}
-	aggregatorServer, err := createAggregatorServer(aggregatorConfig, kubeAPIServer.GenericAPIServer, sharedInformers, wait.NeverStop)
+	aggregatorServer, err := createAggregatorServer(aggregatorConfig, kubeAPIServer.GenericAPIServer, sharedInformers, stopCh)
 	if err != nil {
 		return err
 	}
-	return aggregatorServer.GenericAPIServer.PrepareRun().Run(wait.NeverStop)
+	return aggregatorServer.GenericAPIServer.PrepareRun().Run(stopCh)
```

Comment:
Compare the stop channel logic with https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L247. If the `Run` here fails, the `createAggregatorServer` might leak go routines. We have to use an "internal stop channel" like in the upper link to decouple the error case from the outer stop channel.